File: Ahorn/entities/MoveBlockRedirect.jl
module CommunalHelperMoveBlockRedirect

using ..Ahorn, Maple
using Ahorn.CommunalHelper

@mapdef Entity "CommunalHelper/MoveBlockRedirect" MoveBlockRedirect(
    x::Integer,
    y::Integer,
    width::Integer=Maple.defaultBlockWidth,
    height::Integer=Maple.defaultBlockHeight,
    direction::String="Up",
    fastRedirect::Bool=false,
    deleteBlock::Bool=false,
    oneUse::Bool=false,
    modifier::Number=0.0,
    operation::String="Add",
    reskinFolder::String="",
    overrideColor::String="",
    overrideUsedColor::String="",
)

const placements = Ahorn.PlacementDict(
    "Move Block Redirect (Communal Helper)" => Ahorn.EntityPlacement(
        MoveBlockRedirect,
        "rectangle",
    ),
    "Move Block Redirect (One Use) (Communal Helper)" => Ahorn.EntityPlacement(
        MoveBlockRedirect,
        "rectangle",
        Dict{String,Any}(
            "oneUse" => true,
        ),
    ),
    "Move Block Redirect (Delete Block) (Communal Helper)" => Ahorn.EntityPlacement(
        MoveBlockRedirect,
        "rectangle",
        Dict{String,Any}(
            "deleteBlock" => true,
        ),
    ),
)

const opTypes = ["Add", "Subtract", "Multiply"]

Ahorn.editingOptions(entity::MoveBlockRedirect) = Dict{String, Any}(
    "direction" => Maple.move_block_directions,
    "operation" => opTypes,
)

Ahorn.minimumSize(entity::MoveBlockRedirect) = Maple.defaultBlockWidth, Maple.defaultBlockHeight

Ahorn.resizable(entity::MoveBlockRedirect) = true, true

Ahorn.selection(entity::MoveBlockRedirect) = Ahorn.getEntityRectangle(entity)

const defaultColor = (251, 206, 54, 255) ./ 255
const deleteColor = (204, 37, 65, 255) ./ 255
const fastColor = (41, 195, 47, 255) ./ 255
const slowColor = (28, 91, 179, 255) ./ 255

const clockwise = ["Up", "Right", "Down", "Left"]
function Ahorn.rotated(entity::MoveBlockRedirect, steps::Int)
    dir = get(entity.data, "direction", "Up")
    idx = findall(d -> d == dir, clockwise)[1]
    entity.data["direction"] = clockwise[mod1(idx + steps, 4)]

    return entity
end

const rotations = Dict{String,Float64}(
    "Up" => pi * 1.5,
    "Right" => 0,
    "Down" => pi / 2,
    "Left" => pi,
)

function getRotation(dir::String)
    if haskey(rotations, dir)
        return rotations[dir]
    elseif (fAngle = tryparse(Float64, dir)) !== nothing
        return fAngle
    else
        return 0
    end
end

function getTexturesAndColor(entity::MoveBlockRedirect)
    path = "objects/CommunalHelper/moveBlockRedirect/"
    reskinFolder = String(get(entity.data, "reskinFolder", ""))
    if reskinFolder != ""
        if last(reskinFolder, 1) != "/"
            reskinFolder = reskinFolder * "/"
        end
        path = "objects/" * reskinFolder
    end

    block = path * "block"
    overrideColorHex = String(get(entity.data, "overrideColor", ""))
    overrideColor = (overrideColorHex != "" && length(overrideColorHex) == 6) ? hexToRGBA(overrideColorHex) : nothing
    
    deleteBlock = Bool(get(entity.data, "deleteBlock", false))
    if deleteBlock
        return block, path * "x", overrideColordeleteColor
    end

    operation = get(entity.data, "operation", "Add")
    modifier = abs(get(entity.data, "modifier", 0.0))
    if operation == "Add"
        if modifier != 0.0
            return block, path * "fast", overrideColor === nothing ? fastColor : overrideColor
        end
    elseif operation == "Subtract"
        if modifier != 0
            return block, path * "slow", overrideColor === nothing ? slowColor : overrideColor
        end
    elseif operation == "Multiply"
        if modifier == 0.0
            return block, path * "x", overrideColor === nothing ? deleteColor : overrideColor
        elseif modifier > 1.0
            return block, path * "fast", overrideColor === nothing ? fastColor : overrideColor
        elseif modifier < 1.0
            return block, path * "slow", overrideColor === nothing ? slowColor : overrideColor
        end
    end

    return block, path * "arrow", overrideColor === nothing ? defaultColor : overrideColor
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::MoveBlockRedirect)
    width = Int(get(entity.data, "width", 32))
    height = Int(get(entity.data, "height", 32))

    direction = String(get(entity.data, "direction", "Up"))
    block, sprite, color = getTexturesAndColor(entity)

    tileWidth = ceil(Int, width / 8)
    tileHeight = ceil(Int, height / 8)

    for i in -1:tileWidth, j in -1:tileHeight
        tx = (i == -1) ? 0 : ((i == tileWidth) ? 16 : 8)
        ty = (j == -1) ? 0 : ((j == tileHeight) ? 16 : 8)
        if i == -1 || i == tileWidth || j == -1 || j == tileHeight
            Ahorn.drawImage(ctx, block, i * 8, j * 8, tx, ty, 8, 8, tint=color)
        end
    end

    # finicky
    Ahorn.Cairo.save(ctx)
    Ahorn.Cairo.translate(ctx, width / 2, height / 2)
    Ahorn.Cairo.rotate(ctx, getRotation(direction))
    Ahorn.Cairo.translate(ctx, -8, -8) # sprite always has a 16 x 16 size
    Ahorn.drawImage(ctx, sprite, 0, 0, tint=color)
    Ahorn.Cairo.restore(ctx)
end

end

File: Loenn/entities/misc/move_block_redirect.lua
local drawableSprite = require("structs.drawable_sprite")
local drawableNinePatch = require("structs.drawable_nine_patch")
local enums = require("consts.celeste_enums")
local communalHelper = require("mods").requireFromPlugin("libraries.communal_helper")

local moveBlockRedirect = {}

local operations = {"Add", "Subtract", "Multiply"}
local directions = enums.move_block_directions

moveBlockRedirect.name = "CommunalHelper/MoveBlockRedirect"
moveBlockRedirect.depth = -8500
moveBlockRedirect.minimumSize = {16, 16}
moveBlockRedirect.fieldInformation = {
    direction = {
        options = directions,
        editable = false
    },
    operation = {
        options = operations,
        editable = false
    },
    modifier = {
        minimumValue = 0.0
    },
    overrideColor = {
        fieldType = "color"
    },
    overrideUsedColor = {
        fieldType = "color"
    }
}

moveBlockRedirect.placements = {
    {
        name = "normal",
        data = {
            width = 16,
            height = 16,
            direction = "Up",
            fastRedirect = false,
            deleteBlock = false,
            oneUse = false,
            modifier = 0.0,
            operation = "Add"
        }
    },
    {
        name = "one_use",
        data = {
            width = 16,
            height = 16,
            direction = "Up",
            fastRedirect = false,
            deleteBlock = false,
            oneUse = true,
            modifier = 0.0,
            operation = "Add"
        }
    },
    {
        name = "delete",
        data = {
            width = 16,
            height = 16,
            direction = "Up",
            fastRedirect = false,
            deleteBlock = true,
            oneUse = false,
            modifier = 0.0,
            operation = "Add"
        }
    },
    {
        name = "reskinnable",
        data = {
            width = 16,
            height = 16,
            direction = "Up",
            fastRedirect = false,
            deleteBlock = false,
            oneUse = false,
            modifier = 0.0,
            operation = "Add",
            reskinFolder = "",
            overrideColor = "ffffff",
            overrideUsedColor = "ffffff"
        }
    }
}

local defaultColor = {251 / 255, 206 / 255, 54 / 255, 255 / 255}
local deleteColor = {204 / 255, 37 / 255, 65 / 255, 255 / 255}
local fastColor = {41 / 255, 195 / 255, 47 / 255, 255 / 255}
local slowColor = {28 / 255, 91 / 255, 179 / 255, 255 / 255}

local defaultPath = "objects/CommunalHelper/moveBlockRedirect/"
local frameTexture = "objects/CommunalHelper/moveBlockRedirect/block"

local rotations = {
    ["Up"] = math.pi * 1.5,
    ["Right"] = 0,
    ["Down"] = math.pi / 2,
    ["Left"] = math.pi
}

local function getMoveBlockRedirectThemeData(entity)
    local path = defaultPath
    local reskinFolder = entity.reskinFolder or ""
    if reskinFolder ~= "" then
        if reskinFolder[string.len(reskinFolder)] ~= "/" then
            reskinFolder = reskinFolder * "/"
        end
        path = "objects/" * reskinFolder
    end

    local overrideColorHex = entity.overrideColor or ""

    local deleteBlock = entity.deleteBlock
    if deleteBlock then
        return {icon = path .. "x", color = communalHelper.hexToColor(overrideColorHex, deleteColor)}
    end

    local operation = entity.operation or "Add"
    local modifier = entity.modifier or 0

    if operation == "Add" then
        if modifier ~= 0.0 then
            return {icon = path .. "fast", color = communalHelper.hexToColor(overrideColorHex, fastColor)}
        end
    elseif operation == "Subtract" then
        if modifier ~= 0.0 then
            return {icon = path .. "slow", color = communalHelper.hexToColor(overrideColorHex, slowColor)}
        end
    elseif operation == "Multiply" then
        if modifier == 0.0 then
            return {icon = path .. "x", color = communalHelper.hexToColor(overrideColorHex, deleteColor)}
        elseif modifier > 1.0 then
            return {icon = path .. "fast", color = communalHelper.hexToColor(overrideColorHex, fastColor)}
        elseif modifier < 1.0 then
            return {icon = path .. "slow", color = communalHelper.hexToColor(overrideColorHex, slowColor)}
        end
    end

    return {icon = path .. "arrow", color = communalHelper.hexToColor(overrideColorHex, defaultColor)}
end

function moveBlockRedirect.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 16, entity.height or 16

    local themeData = getMoveBlockRedirectThemeData(entity)

    -- drawable_nine_patch won't apply the given color as of right now (day of plugin implementation)
    -- this commit will fix this problem once it is included in a newer version of Lönn:
    -- https://github.com/CelestialCartographers/Loenn/commit/0389b59e2b0f2a50b50b691eb6141d4f53c0350f
    local ninePatchOptions = {
        mode = "border",
        color = themeData.color
    }
    local frameSprites = drawableNinePatch.fromTexture(frameTexture, ninePatchOptions, x - 8, y - 8, width + 16, height + 16):getDrawableSprite()

    local sprites = frameSprites

    local direction = entity.direction or "Right"
    local rot = rotations[direction]

    local iconSprite = drawableSprite.fromTexture(themeData.icon, {rotation = rot})
    iconSprite:addPosition(x + math.floor(width / 2), y + math.floor(height / 2))
    iconSprite:setColor(themeData.color)
    table.insert(sprites, iconSprite)

    return sprites
end

return moveBlockRedirect
